File: frontend/src/stores/providers.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { providersApi } from '@/api/providers'
import type { Provider, ProviderCreate, ProviderUpdate } from '@/types/models'

export const useProviderStore = defineStore('providers', () => {
  const providers = ref<Provider[]>([])
  const loading = ref(false)
  const currentCliType = ref<string>('claude_code')

  async function fetchProviders(cliType?: string) {
    loading.value = true
    try {
      const type = cliType || currentCliType.value
      const { data } = await providersApi.list(type)
      providers.value = data
    } finally {
      loading.value = false
    }
  }

  function setCliType(cliType: string) {
    currentCliType.value = cliType
  }

  async function createProvider(data: ProviderCreate) {
    const { data: provider } = await providersApi.create(data)
    providers.value.push(provider)
    return provider
  }

  async function updateProvider(id: number, data: ProviderUpdate) {
    const { data: provider } = await providersApi.update(id, data)
    const index = providers.value.findIndex(p => p.id === id)
    if (index !== -1) {
      providers.value[index] = provider
    }
    return provider
  }

  async function deleteProvider(id: number) {
    await providersApi.delete(id)
    providers.value = providers.value.filter(p => p.id !== id)
  }

  async function reorderProviders(ids: number[]) {
    await providersApi.reorder(ids)
    await fetchProviders()
  }

  async function resetFailures(id: number) {
    await providersApi.resetFailures(id)
    await fetchProviders()
  }

  async function unblacklist(id: number) {
    await providersApi.unblacklist(id)
    await fetchProviders()
  }

  return {
    providers,
    loading,
    currentCliType,
    fetchProviders,
    setCliType,
    createProvider,
    updateProvider,
    deleteProvider,
    reorderProviders,
    resetFailures,
    unblacklist
  }
})

File: backend/app/api/v1/providers.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List, Optional

from app.core.database import get_db
from app.schemas.schemas import (
    ProviderCreate, ProviderUpdate, ProviderResponse, ProviderReorder
)
from app.services.provider_service import ProviderService

router = APIRouter()


@router.get("", response_model=List[ProviderResponse])
async def list_providers(cli_type: Optional[str] = None, db: AsyncSession = Depends(get_db)):
    service = ProviderService(db)
    return await service.list_providers(cli_type)


@router.post("", response_model=ProviderResponse)
async def create_provider(data: ProviderCreate, db: AsyncSession = Depends(get_db)):
    service = ProviderService(db)
    return await service.create_provider(data)


@router.get("/{provider_id}", response_model=ProviderResponse)
async def get_provider(provider_id: int, db: AsyncSession = Depends(get_db)):
    service = ProviderService(db)
    provider = await service.get_provider(provider_id)
    if not provider:
        raise HTTPException(status_code=404, detail="Provider not found")
    return provider


@router.put("/{provider_id}", response_model=ProviderResponse)
async def update_provider(provider_id: int, data: ProviderUpdate, db: AsyncSession = Depends(get_db)):
    service = ProviderService(db)
    provider = await service.update_provider(provider_id, data)
    if not provider:
        raise HTTPException(status_code=404, detail="Provider not found")
    return provider


@router.delete("/{provider_id}")
async def delete_provider(provider_id: int, db: AsyncSession = Depends(get_db)):
    service = ProviderService(db)
    success = await service.delete_provider(provider_id)
    if not success:
        raise HTTPException(status_code=404, detail="Provider not found")
    return {"message": "Provider deleted"}


@router.post("/reorder")
async def reorder_providers(data: ProviderReorder, db: AsyncSession = Depends(get_db)):
    service = ProviderService(db)
    await service.reorder_providers(data.ids)
    return {"message": "Providers reordered"}


@router.post("/{provider_id}/reset-failures")
async def reset_failures(provider_id: int, db: AsyncSession = Depends(get_db)):
    service = ProviderService(db)
    success = await service.reset_failures(provider_id)
    if not success:
        raise HTTPException(status_code=404, detail="Provider not found")
    return {"message": "Failures reset"}


@router.post("/{provider_id}/unblacklist")
async def unblacklist_provider(provider_id: int, db: AsyncSession = Depends(get_db)):
    service = ProviderService(db)
    success = await service.unblacklist(provider_id)
    if not success:
        raise HTTPException(status_code=404, detail="Provider not found")
    return {"message": "Provider unblacklisted"}
